crypto/evp/evp_enc.c

```diff
@@ -323,40 +289,42 @@ int EVP_EncryptUpdate(EVP_CIPHER_CTX *ctx, unsigned char *out, int *outl,
         *outl = 0;
         return inl == 0;
     }
-    if (is_partially_overlapping(out + ctx->buf_len, in, cmpl)) {
+
+    i = ctx->buf_len;
+    if (is_partially_overlapping(out, in, cmpl, i, bl)) {
```

Comment:
While this line looks dandy I'd still argue that is_partially_overlapping is unnecessary complex. If I may make suggestion, why not this:
```
bl = ctx->cipher->block_size;
i = ctx->buf_len;
assert(bl > 1 || i == 0);
if ((PTRDIFF_T)(out + i) != (PTRDIRR_T)in    /* in-place? */
    && is_overlapping(out, (i + cmpl) & -bl, in, cmpl)) {
        bad ...
}
```
where is_overlapping is
```
int is_overlapping(PTRDIFF_T ptr1, PTRDIFF_T len1, PTRDIFF_T ptr,2 PTRDIFF_T len2)
{
    PTRDIFF_T diff = ptr2 - ptr1;

    return len2 > 0 & (diff < len1 | (diff > (0 - len2);
}
```
Note that either length of 0 counts as no overlap. Then note that checks for overlapping and in-place are effectively separated. This is to emphasize the difference, and it also improves readability because is_overlapping is *exactly* what intuition tells you. no "secret society knowledge" is required.

Reply:
It depends again on the definition of "partially overlapping".

As I understood the term initially, it would always imply "in!=out && ..."
But to me the exception "out+i != in && ..." is still somehow artificially.
It is only one possible overlapping that works, and the user must have
knowledge of the internal state if the cipher context.

Now I understood we trust our users to know such things.
And we want to support _all or nothing_ with regard to overlapping.

So I try to support another use case where encrypt in-place but
with some separation between the in and out blocks.

I think of encrypting a small header followed by a big payload, 
and again a small trailer.

I realized, that if there is just two cipher blocks room, you need only
one big block that can hold the complete encrypted data.
One can place the payload at the end of the output block,
in a differently partially overlapping way.

Reply:
> As I understood the term initially, it would always imply "in!=out && ..."

Correct. And suggestion is to explicitly rip out the first term of condition and "re-brand" it as "in-place or non-overlapping".

> But to me the exception "out+i != in && ..." is still somehow artificially.

Yeah, it's somewhat counter-intuitive, but it's correct.

> It is only one possible overlapping that works,

Keep in mind that it's counter-intuitive only for block-oriented modes, such as CBC, ECB. Things like CTR, XTS, etc. behave totally intuitive. [Recall that `i` is always zero, as hinted by assert.]

> and the user must have knowledge of the internal state if the cipher context.

Yes, but all user has to do is to use ctx consistently.

> So I try to support another use case where encrypt in-place but
with some separation between the in and out blocks.

I'd say don't, it's not worth the complexity.

Reply:
It's not much complexity, test cases are there for it,
and it used to work in 1.0.1

Reply:
> It's not much complexity

Things add up, and each step can be viewed as "that's not much of additional complexity", but then some day you step back and drop your arms...

> it used to work in 1.0.1

But that was incidental, since there were no checks.

Reply:
But it used to work.

Reply:
I'd like to suggest to solve one problem at a time. So far we effectively established that we support in-place and non-overlapping. If you reckon that block-size-specific partial overlapping is worth supporting, I'd rather prefer it to be discussed on project list first. This effectively means that I'd prefer if this request is resolved without block-size-specific partial overlapping :-) Sorry :-)

Reply:
That's what I wrote on the project list
https://mta.openssl.org/pipermail/openssl-project/2018-March/000346.html

where I wrote this:
"I think for instance of WRAP mode, where the primitive basically
supports arbitrary overlapping (uses memmove) but we can no longer
use that feature because of the partially overlapping check in the
EVP cipher handling.

And the other use case, which I think is worth to mention is

unsigned char *inp = buf + sizeof(buf) - 1, *out = buf;
for (i = 0; i<sizeof(buf)-X; i++) {
     *inp = getc();
     EVP_EncryptUpdate(ctx, out, &outl, inp, 1);
     out += outl;
}"

So I thought everybody should be informed, at least no one objected.

Reply:
As for second use case. I don't quite follow. It works with what I suggest, i.e. wouldn't need special treatment. [We assume that sizeof(buf) is multiple of block size. And it has to be, because of possible padding.] As for WRAP. I admit I have to look closer... But note that it has CUSTOM_CIPHER flag and it's underlying primitive that is responsible for overlap check.

Reply:
Yes, and after I noticed that the current check uses the wrong return value
in the partially overlapping code, I saw that modes/wrap128.c always
does "memmove(out, in + 8, inlen)" so that check can be removed.

Reply:
Just to be clear, I don't feel that "So I try to support another use case where encrypt in-place but with some separation between the in and out blocks" was verbalized in the message clearly enough. And secondly I'd rather insist on explicit feedback than view lack of one as consent.

Reply:
For the second use case, a better pseudo code would be:

```
unsigned char *inp = buf + sizeof(buf) - X, *out = buf;
EVP_EncryptUpdate(ctx, out, &outl, &head, sizeof(head));
out += outl;
EVP_EnctyptUpdate(ctx, out, &outl, in, X);
out += outl;
EVP_EncryptUpdate(ctx, out, &outl, &tail, sizeof(tail));
out += outl;
```

with X > 75% sizeof(buf)
sizeof(head), sizeof(tail) not necessary multiple of block size